Source/Core/Core/ScriptEngine.h

```diff
@@ -0,0 +1,91 @@
+// Copyright 2020 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "Common/CommonTypes.h"
+
+struct lua_State;
+class IniFile;
+
+namespace ScriptEngine
+{
+struct ScriptTarget
+{
+  std::string file_path;
+  bool active = false;
+};
+
+// Top-level script instance definition and state.
+class Script
+{
+public:
+  // Shared class managing all resources held by an instance of a script.
+  class Context
+  {
+  public:
+    explicit Context(lua_State* lua_state) : L(lua_state) {}
+    ~Context();
+    void ExecuteHook(int lua_ref);
+
+  public:
+    Context* self = nullptr;
+
+  private:
+    lua_State* L;  // nil if failed to init
+  };
+
+public:
+  explicit Script(std::string file_path);
+  ~Script();
+
+  // Script is the sole owner of the context, thus copying is forbidden.
+  Script(const Script&) = delete;
+  Script(Script&&) noexcept;
+
+private:
+  void Load();
+  void Unload();
+
+private:
+  // Target state
+  std::string m_file_path;
+
+private:
+  Context* m_ctx = nullptr;  // nullptr if script invalid
+};
+
+// External pointer to a Lua function, for use in Core/PowerPC.
+// Must not be used after unloading the script.
+class LuaFuncHandle
+{
+public:
+  LuaFuncHandle(Script::Context* ctx, const void* lua_ptr, int lua_ref)
+      : m_ctx(ctx), m_lua_ptr(lua_ptr), m_lua_ref(lua_ref)
+  {
+  }
+  inline void Execute() { m_ctx->ExecuteHook(m_lua_ref); };
```

Comment:
inline isn't necessary here, along with the three other functions.

Member functions declared and defined inside a class body are already inline by default.